File: tp2/src/localisation-globale.py
import numpy as np
import math as m
import matplotlib.pyplot as plt
import scipy.io as scio
import shapely.geometry as geo

from particule_resampling import particule_resampling
from gauss import gauss

def main():
    real_map = scio.loadmat("../res/Carte.mat")
    coord_carte = list(zip(real_map['Carte'][0], real_map['Carte'][1]))
    carte = geo.Polygon(coord_carte)

    AZx = list()
    AZy = list()
    ATime = list()

    ### Inserer code de localisation ici
    lidar_direction = [0, m.pi/2, m.pi, (3*m.pi)/2]
    lidar_distance_max = 20
    sig_Lidar = 0.01
    sig_vitesse = 0.01
    sig_omega = 0.05
    sig_compas = 0.01
    sig_angle = 0.01 # bruit à ajouter sur le compas magnétique

    n_step = 100
    dt = 0.1
    n_particules = 100 #nombres de particules
    Reff = 0.1 # Ratio effectif
    vitesse = 0

    Xvrai = np.transpose([[2, 1]])

    x = np.random.rand(n_particules, 1)*18
    y = np.random.rand(n_particules, 1)*8
    X = np.transpose(np.column_stack([x, y]))
    w = [1]*n_particules

    for i_step in range(0, n_step):
        time = i_step*dt
        angle_compas = np.random.normal()

        ### Simulation des vrais mesures ###
        Xvrai[0][0] = Xvrai[0][0] + dt*(vitesse + np.random.normal())*(round(m.cos(angle_compas), 100))
        Xvrai[1][0] = Xvrai[1][0] + dt*(vitesse + np.random.normal())*(round(m.sin(angle_compas), 100))


        # Estimation de la mesure du Lidar
        lidar_estimate = []
        for direction in lidar_direction:
            point_start_line = [Xvrai[0][0], Xvrai[1][0]]
            point_end_line = [point_start_line[0] + lidar_distance_max*(round(m.cos(direction), 100)),
                              point_start_line[1] + lidar_distance_max*(round(m.sin(direction), 100))]

            line = geo.LineString([point_start_line, point_end_line])
            intersection = np.array(carte.intersection(line))
            # Bruit sur le Lidar
            intersection[0] = intersection[0] + np.random.normal()
            intersection[1] = intersection[1] + np.random.normal()

            lidar_estimate.append(intersection)

        ### Fin simulation du monde

        # Filtre a particules
        line_x = geo.LineString([lidar_estimate[0], lidar_estimate[2]])
        line_y = geo.LineString([lidar_estimate[1], lidar_estimate[3]])
        intersection_position = np.array(line_x.intersection(line_y))


        for i_particule in range(0, n_particules):
            X[0][i_particule] = intersection_position[0]
            X[1][i_particule] = intersection_position[1]

            w[i_particule] = gauss(x=angle_compas, sigma=sig_angle) * w[i_particule]

        #X, w = particule_resampling(X=X, w=w, ratio=Reff)

        # ========= Fin des equations du filtre EKF ou particule =============

        # Cueillette des donnees pour les graphiques/statistiques
        AZx.append(intersection_position[0])
        AZy.append(intersection_position[1])
        ATime.append(time)


    myFontSize = 14 # Taille de la police de caractere pour les graphes
    plt.clf()

    plt.plot(real_map['Carte'][0], real_map['Carte'][1])
    plt.plot(np.transpose(X[0]), np.transpose(X[1]), 'b*', label = 'Particules au depart')
    #plt.plot(ATime, AX1, 'go', label='filtre a particules')
    plt.plot(AZx, AZy, 'r*', label='Mesure h_z^{-1}') # Ici on peut inverser le capteur, pour trouver la position correspondant a z.

    plt.xlabel('Temps (s)')
    plt.ylabel('Estime de position (m)')
    plt.legend()

    plt.draw()
    plt.show()


if __name__ == "__main__":
    main()

File: tp2/LocalisationGlobale.m
clearvars
load('res/Carte.mat')

myFontSize = 14; % Taille de la police de caractere pour les graphes
nStep = 100;
dt = 0.1;
nParticules = 100;
Reff = 0.5;

lidarDirection = [0, (pi/2), pi, ((3*pi)/2)];
lidarDistanceMax = 20;
sigLidar = 0.01;
sigVitesse = 0.01;

vitesse = 0;
angleRobot = 0;
sigAngle = 0.01;
sigPositionEstime = 1;
sigVitesseEstime = sigVitesse*;

%x = rand(1, nParticules)*18;
%y = rand(1, nParticules)*8;

x = ones(1, nParticules);
y = 2*ones(1, nParticules);

Xvrai = [1; 2];
X = [x; y];

w = ones(1, nParticules)./nParticules;


X2Vrai = [];
X2Mesure = [];

for iStep = 1:nStep
    time = iStep*dt;
    angleCompas = angleRobot + sigAngle*randn;

    %%% Simulation des mesures %%%
    Xvrai(1) = Xvrai(1) + dt*(vitesse + sigVitesse*randn)*cos(angleCompas);
    Xvrai(2) = Xvrai(2) + dt*(vitesse + sigVitesse*randn)*sin(angleCompas);
    
    [intersections_x, intersections_y] = ClosestLidarIntersection(Carte, Xvrai, angleCompas, sigLidar);
        
    ligne1_x = [intersections_x(1), intersections_x(3)];
    ligne1_y = [intersections_y(1), intersections_y(3)];
    ligne2_x = [intersections_x(2), intersections_x(4)];
    ligne2_y = [intersections_y(2), intersections_y(4)];
    
    [Xmesure, Ymesure] = polyxpoly(ligne1_x, ligne1_y, ligne2_x, ligne2_y);
    
    
    %%% Filtre a particules
    for iParticule=1:nParticules
        X(1, iParticule) = X(1, iParticule) + dt*(vitesse + sigVitesseEstime*randn*10)*cos(angleCompas);
        X(2, iParticule) = X(2, iParticule) + dt*(vitesse + sigVitesseEstime*randn*10)*sin(angleCompas);
        wnew = 0;
        [estime_intersection_x, estime_intersection_y] = ClosestLidarIntersection(Carte, X(:, iParticule), angleCompas, sigLidarEstime);
        
        for iDirection = 1:length(lidarDirection)
            dis_x = estime_intersection_x(iDirection) - intersections_x(iDirection);
            dis_y = estime_intersection_y(iDirection) - intersections_y(iDirection);
            wnew = wnew*gauss(dis_x, sigLidarEstime);
            wnew = wnew*gauss(dis_y, sigLidarEstime);
        end
       display(wnew);
       
        w(iParticule) = w(iParticule)*wnew;
    end
    [X, w] = ParticuleResampling(X, w, Reff);
    
    %%% Dessin des données
    X2Vrai = [X2Vrai Xvrai];
    %X2Mesure = [X2Mesure [Xmesure; Ymesure]];
    clf
    h(1) = plot(Carte(1, :), Carte(2, :));
    hold on
    PlotParticules2D(X,angleRobot, w,'o',0.1);
    h(2) = plot(intersections_x', intersections_y', 'r*');
    h(3) = plot(Xvrai(1), Xvrai(2), 'k+');
    %h(5) = plot(X2Mesure(1, :), X2Mesure(2, :), 'go');
    hold off
    xlabel('Temps (s)');
    ylabel('Estime de position (m)');
    legend(h,{'Carte reel', 'lidar', 'position reel'});

    drawnow();
end